pkg/mimirpb/custom.go

```diff
@@ -38,6 +44,183 @@ func (m *WriteRequest) MinTimestamp() int64 {
+func (m *WriteRequest) SizeWithoutMetadata() int {
+	if len(m.Metadata) == 0 {
+		return m.Size()
+	}
+
+	origMetadata := m.Metadata
+	m.Metadata = nil
+	size := m.Size()
+	m.Metadata = origMetadata
+
+	return size
+}
+
+// SizeWithoutTimeseries is like Size() but subtracts the marshalled Timeseries size.
+func (m *WriteRequest) SizeWithoutTimeseries() int {
+	if len(m.Timeseries) == 0 {
+		return m.Size()
+	}
+
+	origTimeseries := m.Timeseries
+	m.Timeseries = nil
+	size := m.Size()
+	m.Timeseries = origTimeseries
+
+	return size
+}
+
+// SplitByMaxMarshalSize splits the WriteRequest into multiple ones, where each partial WriteRequest marshalled size
+// is at most maxSize.
+//
+// This function guarantees that a single Timeseries or Metadata entry is never split across multiple requests.
+// For this reason, this function is a best-effort: if a single Timeseries or Metadata marshalled size is bigger
+// than maxSize, then the returned partial WriteRequest marshalled size will be bigger than maxSize too.
+//
+// The returned partial WriteRequests are NOT a deep copy of the input one; they may contain references to slices
+// and data from the original WriteRequest.
+// TODO benchmark
+// TODO unit test: consider adding a fuzzy test to stress this logic
+func (m *WriteRequest) SplitByMaxMarshalSize(maxSize int) []*WriteRequest {
+	if m.Size() <= maxSize {
+		return []*WriteRequest{m}
+	}
+
+	partialReqsWithTimeseries := m.splitTimeseriesByMaxMarshalSize(maxSize)
+	partialReqsWithMetadata := m.splitMetadataByMaxMarshalSize(maxSize)
+
+	// Most of the times a write request only have either Timeseries or Metadata.
+	if len(partialReqsWithMetadata) == 0 {
+		return partialReqsWithTimeseries
+	}
+	if len(partialReqsWithTimeseries) == 0 {
+		return partialReqsWithMetadata
+	}
+
+	merged := make([]*WriteRequest, 0, len(partialReqsWithTimeseries)+len(partialReqsWithMetadata))
+	merged = append(merged, partialReqsWithTimeseries...)
+	merged = append(merged, partialReqsWithMetadata...)
+
+	return merged
+}
+
+func (m *WriteRequest) splitTimeseriesByMaxMarshalSize(maxSize int) []*WriteRequest {
+	if len(m.Timeseries) == 0 {
+		return nil
+	}
+
+	newPartialReq := func(preallocTimeseries int) *WriteRequest {
+		return &WriteRequest{
+			Timeseries:              preallocSliceIfNeeded[PreallocTimeseries](preallocTimeseries),
+			Source:                  m.Source,
+			SkipLabelNameValidation: m.SkipLabelNameValidation,
+		}
+	}
+
+	// The partial requests returned by this function will not contain any Metadata,
+	// so we first compute the request size without it.
+	reqSizeWithoutMetadata := m.SizeWithoutMetadata()
+	if reqSizeWithoutMetadata <= maxSize {
+		partialReq := newPartialReq(0)
+		partialReq.Timeseries = m.Timeseries
+		return []*WriteRequest{partialReq}
+	}
+
+	// We assume that different timeseries roughly have the same size (no huge outliers)
+	// so we preallocate the returned slice just adding 1 extra item (+2 because a +1 is to round up).
```

Comment:
I could understand +1, but why +2 again?

Reply:
+1 to round up, and +1 for an extra item. Does this answer your question?